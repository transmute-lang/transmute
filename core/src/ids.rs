macro_rules! make_id {
    ($name:ident) => {
        #[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Default)]
        pub struct $name {
            id: usize,
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.id)
            }
        }

        impl std::fmt::Debug for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}({})", stringify!($name), self.id)
            }
        }

        impl From<usize> for $name {
            fn from(id: usize) -> Self {
                Self { id }
            }
        }

        impl From<$name> for usize {
            fn from(id: $name) -> Self {
                id.id
            }
        }

        impl std::ops::Add<usize> for $name {
            type Output = $name;

            fn add(self, rhs: usize) -> Self::Output {
                Self::from(self.id + rhs)
            }
        }

        impl std::ops::Add<usize> for &$name {
            type Output = $name;

            fn add(self, rhs: usize) -> Self::Output {
                $name::from(self.id + rhs)
            }
        }
    };
}

pub(crate) use make_id;

make_id!(IdentId);
make_id!(IdentRefId);
make_id!(ExprId);
make_id!(StmtId);
make_id!(SymbolId);
make_id!(TypeId);
make_id!(FunctionId);
make_id!(StructId);
make_id!(TypeDefId);

#[macro_export]
macro_rules! id {
    ($e:expr) => {
        <_ as Into<usize>>::into($e)
    };
}

pub use id;
