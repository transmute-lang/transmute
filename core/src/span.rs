use std::cmp::Ordering;
use std::fmt::{Debug, Formatter};

#[derive(Default, Clone, Eq)]
pub struct Span {
    pub line: usize,
    pub column: usize,
    pub start: usize,
    pub len: usize,
}

impl Span {
    pub fn new(line: usize, column: usize, start: usize, len: usize) -> Self {
        assert!(line >= 1);
        assert!(column >= 1);
        Self {
            line,
            column,
            start,
            len,
        }
    }

    pub fn extend_to(&self, other: &Span) -> Self {
        let self_end = self.start + self.len;
        let other_end = other.start + other.len;
        let diff = other_end - self_end;
        Self {
            line: self.line,
            column: self.column,
            start: self.start,
            len: self.len + diff,
        }
    }

    pub fn extend(&self, len: usize) -> Self {
        Self {
            line: self.line,
            column: self.column,
            start: self.start,
            len: self.len + len,
        }
    }

    pub fn end(&self) -> usize {
        self.start + self.len
    }
}

impl PartialEq<Self> for Span {
    fn eq(&self, other: &Self) -> bool {
        self.start == other.start && self.len == other.len
    }
}

impl PartialOrd<Self> for Span {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Span {
    fn cmp(&self, other: &Self) -> Ordering {
        if self == other {
            return Ordering::Equal;
        }
        if self.start < other.start {
            return Ordering::Less;
        }
        if self.start == other.start && self.len < other.len {
            return Ordering::Less;
        }
        Ordering::Greater
    }
}

impl Debug for Span {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}:{}; [{}, {}]",
            self.line,
            self.column,
            self.start,
            self.start + self.len
        )
    }
}
