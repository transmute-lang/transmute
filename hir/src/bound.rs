use std::fmt::{Debug, Formatter};
use transmute_core::ids::SymbolId;

pub trait BoundState {}

#[derive(Debug, Clone, PartialEq)]
pub struct Unbound;

impl BoundState for Unbound {}

#[derive(Clone, PartialEq)]
pub struct Bound(pub SymbolId);

impl BoundState for Bound {}

impl Debug for Bound {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Bound({})", self.0)
    }
}

#[derive(PartialEq)]
pub struct BoundMultiple(pub Vec<SymbolId>);

impl BoundState for BoundMultiple {}

impl Debug for BoundMultiple {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Bound({:?})", self.0)
    }
}
