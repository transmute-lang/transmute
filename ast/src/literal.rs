use std::fmt::{Debug, Formatter};
use transmute_core::ids::IdentRefId;
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    pub kind: LiteralKind,
    pub span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
