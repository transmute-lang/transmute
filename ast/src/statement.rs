use crate::annotation::Annotation;
use crate::identifier::Identifier;
use transmute_core::ids::{ExprId, IdentRefId, InputId, StmtId, TypeDefId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Statement {
    pub id: StmtId,
    pub kind: StatementKind,
    pub span: Span,
}

impl Statement {
    pub fn new(id: StmtId, kind: StatementKind, span: Span) -> Self {
        Self { id, kind, span }
    }

    pub fn as_namespace(&self) -> (&Identifier, InputId, &[StmtId]) {
        match &self.kind {
            StatementKind::Namespace(identifier, input_id, stmts) => (identifier, *input_id, stmts),
            _ => panic!("{:?} is not a namespace", self),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum StatementKind {
    Expression(ExprId),
    Let(Identifier, ExprId),
    Ret(Option<ExprId>, RetMode),
    LetFn(Identifier, Vec<Annotation>, Vec<Parameter>, Return, ExprId),
    Struct(Identifier, Vec<Annotation>, Vec<Field>),
    Annotation(Identifier),
    Use(Vec<IdentRefId>),
    Namespace(
        /// Namespace identifier
        Identifier,
        /// `InputId` this namespace is coming from
        InputId,
        /// Statements included in this namespace
        Vec<StmtId>,
    ),
}

/// AST only produces `Ret` with `RetMode::Explicit` but having that information in the AST nodes
/// helps in the hir transformations (see `ImplicitRetConverter::visit_expression`).
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RetMode {
    Explicit,
    Implicit,
}

impl RetMode {
    pub fn as_str(&self) -> &'static str {
        match self {
            RetMode::Explicit => "explicit",
            RetMode::Implicit => "implicit",
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypeDef {
    pub kind: TypeDefKind,
    pub span: Span,
}

impl TypeDef {
    pub fn identifier_ref_id(&self) -> &[IdentRefId] {
        match &self.kind {
            TypeDefKind::Simple(ident_ref_id) => ident_ref_id,
            TypeDefKind::Array(_, _) => todo!(),
            _ => panic!("no identifier_ref_id for TypeDefKind::Dummy"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeDefKind {
    Simple(Vec<IdentRefId>),
    Array(TypeDefId, usize),
    /// A dummy type kind, inserted by the parser when the type could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Parameter {
    pub identifier: Identifier,
    pub type_def_id: TypeDefId,
    pub span: Span,
}

impl Parameter {
    pub fn new(identifier: Identifier, type_def_id: TypeDefId, span: Span) -> Self {
        Self {
            identifier,
            type_def_id,
            span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Return {
    pub type_def_id: Option<TypeDefId>,
}

impl Return {
    pub fn none() -> Self {
        Self { type_def_id: None }
    }

    pub fn some(type_def_id: TypeDefId) -> Self {
        Self {
            type_def_id: Some(type_def_id),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field {
    pub identifier: Identifier,
    pub type_def_id: TypeDefId,
    pub span: Span,
}

impl Field {
    pub fn new(identifier: Identifier, type_def_id: TypeDefId, span: Span) -> Self {
        Self {
            identifier,
            type_def_id,
            span,
        }
    }
}
