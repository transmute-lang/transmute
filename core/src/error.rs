use crate::input::Inputs;
use crate::span::Span;
use std::fmt::{Debug, Display, Formatter};

#[derive(Debug, Default, PartialEq)]
pub struct Diagnostics<I> {
    diagnostics: Vec<Diagnostic>,
    inputs: I,
}

impl<I> Diagnostics<I> {
    pub fn len(&self) -> usize {
        self.diagnostics.len()
    }

    pub fn is_empty(&self) -> bool {
        self.diagnostics.is_empty()
    }

    pub fn iter(&self) -> DiagnosticsIterator<I> {
        DiagnosticsIterator {
            diagnostics: self,
            index: 0,
        }
    }

    pub fn append(&mut self, mut other: Self) {
        self.diagnostics.append(&mut other.diagnostics)
    }

    pub fn push(&mut self, diagnostic: Diagnostic) {
        self.diagnostics.push(diagnostic);
    }

    pub fn report_err<S: Into<String>>(
        &mut self,
        message: S,
        span: Span,
        generated_at: (&'static str, u32),
    ) {
        self.push(Diagnostic::new(message, span, Level::Error, generated_at));
    }
}

impl Diagnostics<()> {
    pub fn with_inputs(self, inputs: Inputs) -> Diagnostics<Inputs> {
        Diagnostics {
            diagnostics: self.diagnostics,
            inputs,
        }
    }
}

impl Display for Diagnostics<()> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut diagnostics = self.diagnostics.to_vec();
        diagnostics.sort_by(|a, b| a.span.cmp(&b.span));

        for diagnostic in diagnostics {
            writeln!(f, " - {}", diagnostic)?;
        }
        Ok(())
    }
}

impl Display for Diagnostics<Inputs> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut diagnostics = self.diagnostics.to_vec();
        diagnostics.sort_by(|a, b| a.span.cmp(&b.span));

        for diagnostic in diagnostics {
            writeln!(f, " - {}", diagnostic.format(&self.inputs))?;
        }
        Ok(())
    }
}

pub struct DiagnosticsIterator<'a, I> {
    diagnostics: &'a Diagnostics<I>,
    index: usize,
}

impl<'a, I> Iterator for DiagnosticsIterator<'a, I> {
    type Item = &'a Diagnostic;

    fn next(&mut self) -> Option<Self::Item> {
        if self.index < self.diagnostics.len() {
            self.index += 1;
            Some(&self.diagnostics.diagnostics[self.index - 1])
        } else {
            None
        }
    }
}

#[derive(Clone, PartialEq)]
pub struct Diagnostic {
    pub message: String,
    pub span: Span,
    pub level: Level,
    pub generated_at: (&'static str, u32),
}

impl Diagnostic {
    pub fn format(&self, inputs: &Inputs) -> String {
        #[cfg(debug_assertions)]
        {
            format!(
                "[{generated_at_file}:{generated_at_line}] {source}:{line}:{column}: {message}",
                generated_at_file = self.generated_at.0,
                generated_at_line = self.generated_at.1,
                message = self.message,
                source = inputs[self.span.input_id].name(),
                line = self.span.line,
                column = self.span.column
            )
        }
        #[cfg(not(debug_assertions))]
        {
            format!(
                "{}:{}:{}: {}",
                self.message,
                inputs[self.span.input_id].name(),
                self.span.line,
                self.span.column
            )
        }
    }
}

impl Display for Diagnostic {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        #[cfg(debug_assertions)]
        {
            write!(
                f,
                "[{generated_at_file}:{generated_at_line}] {line}:{column}: {message}",
                generated_at_file = self.generated_at.0,
                generated_at_line = self.generated_at.1,
                message = self.message,
                line = self.span.line,
                column = self.span.column
            )
        }
        #[cfg(not(debug_assertions))]
        {
            write!(
                f,
                "{}:{}: {}",
                self.span.line, self.span.column, self.message
            )
        }
    }
}

impl Debug for Diagnostic {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Diagnostic {{")?;
        writeln!(f, "    message: {:?},", self.message)?;
        writeln!(f, "    span: {:?},", self.span)?;
        writeln!(f, "    level: {:?},", self.level)?;
        #[cfg(feature = "print_generated_at_in_diagnostic_debug")]
        {
            writeln!(
                f,
                "    generated_at: {}:{},",
                self.generated_at.0, self.generated_at.1
            )?;
        }
        write!(f, "}}")
    }
}

impl Diagnostic {
    pub fn new<S: Into<String>>(
        message: S,
        span: Span,
        level: Level,
        generated_at: (&'static str, u32),
    ) -> Self {
        Self {
            message: message.into(),
            span,
            level,
            generated_at,
        }
    }

    pub fn error<S: Into<String>>(
        message: S,
        span: Span,
        generated_at: (&'static str, u32),
    ) -> Self {
        Self::new(message, span, Level::Error, generated_at)
    }
}

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
pub enum Level {
    Error,
}
