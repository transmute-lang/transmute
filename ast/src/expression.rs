use crate::literal::Literal;
use crate::operators::{BinaryOperator, UnaryOperator};
use std::fmt::Debug;
use transmute_core::ids::{ExprId, IdentRefId, StmtId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression {
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
}

impl Expression {
    pub fn new(id: ExprId, kind: ExpressionKind, span: Span) -> Self {
        Self { id, kind, span }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Binary(ExprId, BinaryOperator, ExprId),
    Unary(UnaryOperator, ExprId),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    /// A dummy expression kind, inserted by the parser when the expression could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}
