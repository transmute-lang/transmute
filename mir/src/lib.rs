#![allow(unused)]

use std::collections::HashMap;
use std::env::var;
use std::thread::available_parallelism;
use transmute_core::error::{Diagnostic, Diagnostics, Level};
use transmute_core::id;
use transmute_core::ids::{
    ExprId, FunctionId, IdentId, IdentRefId, StmtId, StructId, SymbolId, TypeId,
};
use transmute_core::span::Span;
use transmute_core::vec_map::VecMap;
use transmute_hir::bound::{Bound, BoundState};
use transmute_hir::expression::{ExpressionKind as HirExpressionKind, Target as HirTarget};
use transmute_hir::literal::{Literal as HirLiteral, LiteralKind as HirLiteralKind};
use transmute_hir::natives::Type as HirType;
use transmute_hir::statement::{Field, RetMode, Return, StatementKind as HirStatementKind};
use transmute_hir::symbol::Symbol;
use transmute_hir::typed::{Typed, TypedState};
use transmute_hir::{
    ResolvedExpression as HirExpression, ResolvedIdentifier as HirIdentifier,
    ResolvedParameter as HirParameter, ResolvedStatement as HirStatement,
    ResolvedIdentifierRef as HirIdentifierRef
};
use transmute_hir::{ResolvedHir as Hir, ResolvedReturn as HirReturn};

pub fn make_mir(hir: Hir) -> Result<Mir, Diagnostics> {
    Mir::try_from(hir)
}

struct Transformer {
    functions: VecMap<FunctionId, Function>,
    structs: VecMap<StructId, Struct>,
    expressions: VecMap<ExprId, Expression>,
    statements: VecMap<StmtId, Statement>,
}

impl Transformer {
    pub fn new() -> Self {
        Self {
            functions: Default::default(),
            structs: Default::default(),
            expressions: Default::default(),
            statements: Default::default(),
        }
    }

    pub fn transform(mut self, mut hir: Hir) -> Result<Mir, Diagnostics> {
        let mut diagnostics = Diagnostics::default();

        self.expressions.resize(hir.expressions.len());
        self.statements.resize(hir.statements.len());

        for stmt_id in hir.roots.clone().iter() {
            let stmt = hir.statements.remove(*stmt_id).unwrap();

            match stmt.kind {
                HirStatementKind::LetFn(identifier, parameters, ret_type, expr_id) => {
                    self.transform_function(&mut hir, identifier, parameters, ret_type, expr_id)
                }
                HirStatementKind::Struct(_, _) => todo!(),
                _ => {
                    // todo it seems that roots is an useless concept outside of the interpreter flow. think
                    //   about what to do about it
                    diagnostics.push(Diagnostic::new(
                        "only functions are allowed at top level",
                        stmt.span,
                        Level::Error,
                        (file!(), line!()),
                    ))
                }
            }
        }

        if !diagnostics.is_empty() {
            return Err(diagnostics);
        }

        Ok(Mir {
            functions: self.functions,
            structs: self.structs,
            identifiers: hir.identifiers,
            identifier_refs: hir.identifier_refs.into_iter().map(|(ident_ref_id, ident_ref)| {
                (ident_ref_id, IdentifierRef::from(ident_ref))
            }).collect::<VecMap<IdentRefId, IdentifierRef>>(),
            expressions: self.expressions,
            statements: self.statements,
            symbols: hir.symbols,
            types: hir
                .types
                .into_iter()
                .filter_map(|(type_id, ty)| match ty {
                    HirType::Invalid => None,
                    HirType::Boolean => Some((type_id, Type::Boolean)),
                    HirType::Function(params, ret_type) => {
                        Some((type_id, Type::Function(params, ret_type)))
                    }
                    HirType::Struct(_) => Some((type_id, todo!())),
                    HirType::Number => Some((type_id, Type::Number)),
                    HirType::Void => Some((type_id, Type::Void)),
                    HirType::None => Some((type_id, Type::None)),
                })
                .collect::<VecMap<TypeId, Type>>(),
        })
    }

    fn transform_function(
        &mut self,
        hir: &mut Hir,
        identifier: HirIdentifier,
        parameters: Vec<HirParameter>,
        ret: HirReturn,
        expr_id: ExprId,
    ) {
        let function_id = FunctionId::from(self.functions.len());

        let expression = hir.expressions.remove(expr_id).unwrap();
        let (mutated_symbol_ids, mut variables) = self.transform_expression(hir, expression);

        let parameters = parameters
            .into_iter()
            .map(|param| Parameter {
                symbol_id: param.resolved_symobl_id(),
                type_id: param.resolved_type_id(),
                mutable: mutated_symbol_ids.contains(&param.resolved_symobl_id()),
                identifier: param.identifier.into(),
            })
            .collect::<Vec<Parameter>>();

        for variable in variables.values_mut() {
            if mutated_symbol_ids.contains(&variable.symbol_id) {
                variable.mutable = true
            }
        }

        let function = Function {
            symbol_id: identifier.resolved_symbol_id(),
            identifier: identifier.into(),
            parameters,
            variables,
            body: expr_id,
            ret: ret.resolved_type_id(),
        };

        self.functions.insert(function_id, function);
    }

    fn transform_expression(
        &mut self,
        hir: &mut Hir,
        expression: HirExpression,
    ) -> (Vec<SymbolId>, HashMap<SymbolId, Variable>) {
        let type_id = expression.resolved_type_id();

        match expression.kind {
            HirExpressionKind::Assignment(target, expr_id) => self.transform_assignment(
                hir,
                expression.id,
                expression.span,
                type_id,
                target,
                expr_id,
            ),
            HirExpressionKind::If(_, _, _) => todo!(),
            HirExpressionKind::Literal(literal) => {
                self.transform_literal(hir, expression.id, expression.span, type_id, literal);
                (Default::default(), Default::default())
            }
            HirExpressionKind::Access(_, _) => todo!(),
            HirExpressionKind::FunctionCall(ident_ref_id, params) => self.transform_function_call(
                hir,
                expression.id,
                expression.span,
                type_id,
                ident_ref_id,
                params,
            ),
            HirExpressionKind::While(_, _) => todo!(),
            HirExpressionKind::Block(stmt_ids) => {
                self.transform_block(hir, expression.id, expression.span, type_id, stmt_ids)
            }
            HirExpressionKind::StructInstantiation(_, _) => todo!(),
        }
    }

    fn transform_literal(
        &mut self,
        hir: &mut Hir,
        expr_id: ExprId,
        span: Span,
        type_id: TypeId,
        literal: HirLiteral,
    ) {
        match literal.kind {
            HirLiteralKind::Boolean(b) => self.expressions.insert(
                expr_id,
                Expression {
                    id: expr_id,
                    kind: ExpressionKind::Literal(Literal {
                        kind: LiteralKind::Boolean(b),
                        span: literal.span,
                    }),
                    span,
                    type_id,
                },
            ),
            HirLiteralKind::Identifier(ident_ref_id) => self.expressions.insert(
                expr_id,
                Expression {
                    id: expr_id,
                    kind: ExpressionKind::Literal(Literal {
                        kind: LiteralKind::Identifier(
                            hir.identifier_refs[ident_ref_id].resolved_symbol_id(),
                        ),
                        span: literal.span,
                    }),
                    span,
                    type_id,
                },
            ),
            HirLiteralKind::Number(n) => self.expressions.insert(
                expr_id,
                Expression {
                    id: expr_id,
                    kind: ExpressionKind::Literal(Literal {
                        kind: LiteralKind::Number(n),
                        span: literal.span,
                    }),
                    span,
                    type_id,
                },
            ),
        };
    }

    /// Transform the assignment and return the mutated `SymbolId`
    fn transform_assignment(
        &mut self,
        hir: &mut Hir,
        expr_id: ExprId,
        span: Span,
        type_id: TypeId,
        target: HirTarget,
        value_expr_id: ExprId,
    ) -> (Vec<SymbolId>, HashMap<SymbolId, Variable>) {
        let expression = hir.expressions.remove(value_expr_id).unwrap();

        let (mut mutated_symbol_ids, variables) = self.transform_expression(hir, expression);

        match target {
            HirTarget::Direct(ident_ref_id) => {
                let symbol_id = hir.identifier_refs[ident_ref_id].resolved_symbol_id();
                self.expressions.insert(
                    expr_id,
                    Expression {
                        id: expr_id,
                        kind: ExpressionKind::Assignment(Target::Direct(symbol_id), value_expr_id),
                        span,
                        type_id,
                    },
                );

                mutated_symbol_ids.push(symbol_id);
                (mutated_symbol_ids, variables)
            }
            HirTarget::Indirect(_) => todo!(),
        }
    }

    fn transform_function_call(
        &mut self,
        hir: &mut Hir,
        expr_id: ExprId,
        span: Span,
        type_id: TypeId,
        ident_ref_id: IdentRefId,
        params: Vec<ExprId>,
    ) -> (Vec<SymbolId>, HashMap<SymbolId, Variable>) {
        let mut mutated_symbol_ids = Vec::new();
        let mut variables = HashMap::new();

        let symbol_id = hir.identifier_refs[ident_ref_id].resolved_symbol_id();

        for expr_id in params.iter() {
            let expression = hir.expressions.remove(*expr_id).unwrap();
            let (mut new_mutated_symbol_ids, mut new_variables) =
                self.transform_expression(hir, expression);
            mutated_symbol_ids.append(&mut new_mutated_symbol_ids);
            variables.extend(new_variables);
        }

        self.expressions.insert(
            expr_id,
            Expression {
                id: expr_id,
                kind: ExpressionKind::FunctionCall(symbol_id, params),
                span,
                type_id,
            },
        );

        (mutated_symbol_ids, variables)
    }

    fn transform_block(
        &mut self,
        hir: &mut Hir,
        expr_id: ExprId,
        span: Span,
        type_id: TypeId,
        stmt_ids: Vec<StmtId>,
    ) -> (Vec<SymbolId>, HashMap<SymbolId, Variable>) {
        let mut kept_stmt_ids = Vec::with_capacity(stmt_ids.len());
        let mut mutated_symbols = Vec::new();
        let mut variables = HashMap::new();

        for stmt_id in stmt_ids {
            let stmt = hir.statements.remove(stmt_id).unwrap();
            match stmt.kind {
                HirStatementKind::Expression(_) => todo!(),
                HirStatementKind::Let(ident, expr_id) => {
                    let variable = self.transform_let(hir, ident, expr_id);
                    variables.insert(variable.symbol_id, variable);
                }
                HirStatementKind::Ret(expr_id, _) => {
                    mutated_symbols
                        .append(&mut self.transform_ret(hir, stmt.id, stmt.span, expr_id));
                    kept_stmt_ids.push(stmt.id)
                }
                HirStatementKind::LetFn(_, _, _, _) => todo!(),
                HirStatementKind::Struct(_, _) => todo!(),
            }
        }

        self.expressions.insert(
            expr_id,
            Expression {
                id: expr_id,
                kind: ExpressionKind::Block(kept_stmt_ids),
                span,
                type_id,
            },
        );

        (mutated_symbols, variables)
    }

    fn transform_let(
        &mut self,
        hir: &mut Hir,
        identifier: HirIdentifier,
        expr_id: ExprId,
    ) -> Variable {
        let expression = hir.expressions.remove(expr_id).unwrap();
        let type_id = expression.resolved_type_id();
        let (mutated_symbol_ids, variables) = self.transform_expression(hir, expression);

        assert!(mutated_symbol_ids.is_empty());
        assert!(variables.is_empty());

        Variable {
            symbol_id: identifier.resolved_symbol_id(),
            identifier: identifier.into(),
            type_id,
            expression: expr_id,
            mutable: false, // todo add `mut` keyword?
        }
    }

    /// Transform ret and return the list of mutated `SymbolId`s
    fn transform_ret(
        &mut self,
        hir: &mut Hir,
        stmt_id: StmtId,
        span: Span,
        expr_id: ExprId,
    ) -> Vec<SymbolId> {
        let expr = hir.expressions.remove(expr_id).unwrap();
        let (mutated_symbols, variables) = self.transform_expression(hir, expr);

        // debug_assert!(mutated_symbols.is_empty());
        debug_assert!(variables.is_empty());

        self.statements.insert(
            stmt_id,
            Statement {
                id: stmt_id,
                kind: StatementKind::Ret(expr_id),
                span,
            },
        );

        mutated_symbols
    }
}

#[derive(Debug)]
pub struct Mir {
    pub functions: VecMap<FunctionId, Function>,
    pub structs: VecMap<StructId, Struct>,
    pub identifiers: VecMap<IdentId, String>,
    // todo remove once we have what we need in symbols so the LLVM pass can get the identifier
    //  from the symbol id
    pub identifier_refs: VecMap<IdentRefId, IdentifierRef>,
    pub expressions: VecMap<ExprId, Expression>,
    pub statements: VecMap<StmtId, Statement>,
    pub symbols: VecMap<SymbolId, Symbol>,
    pub types: VecMap<TypeId, Type>,
}

impl TryFrom<Hir> for Mir {
    type Error = Diagnostics;

    fn try_from(hir: Hir) -> Result<Self, Self::Error> {
        Transformer::new().transform(hir)
    }
}

#[derive(Debug)]
pub struct Function {
    pub identifier: Identifier,
    pub symbol_id: SymbolId,
    pub parameters: Vec<Parameter>,
    pub variables: HashMap<SymbolId, Variable>,
    pub body: ExprId,
    pub ret: TypeId,
}

#[derive(Debug)]
pub struct Identifier {
    pub id: IdentId,
    pub span: Span,
}

impl From<HirIdentifier> for Identifier {
    fn from(hir: HirIdentifier) -> Self {
        Identifier {
            id: hir.id,
            span: hir.span,
        }
    }
}

#[derive(Debug)]
pub struct Parameter {
    pub identifier: Identifier,
    pub symbol_id: SymbolId,
    pub type_id: TypeId,
    pub mutable: bool,
}

#[derive(Debug)]
pub struct Variable {
    pub identifier: Identifier,
    pub symbol_id: SymbolId,
    pub type_id: TypeId,
    pub expression: ExprId,
    pub mutable: bool,
}

#[derive(Debug)]
pub struct Struct {
    // todo
}

#[derive(Debug)]
pub struct IdentifierRef {
    pub id: IdentRefId,
    pub ident: Identifier,
}

impl From<HirIdentifierRef>  for IdentifierRef {
    fn from(identifier_ref: HirIdentifierRef) -> Self {
        IdentifierRef {
            id: identifier_ref.id,
            ident: Identifier::from(identifier_ref.ident),
        }
    }
}

#[derive(Debug)]
pub struct Expression {
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
    pub type_id: TypeId,
}

#[derive(Debug)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Access(ExprId, SymbolId),
    FunctionCall(SymbolId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(SymbolId, Vec<(SymbolId, ExprId)>),
}

#[derive(Debug)]
pub struct Literal {
    pub kind: LiteralKind,
    pub span: Span,
}

#[derive(Debug)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(SymbolId),
    Number(i64),
}

#[derive(Debug)]
pub enum Target {
    Direct(SymbolId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}

#[derive(Debug)]
pub struct Statement {
    pub id: StmtId,
    pub kind: StatementKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub enum StatementKind {
    Expression(ExprId),
    Ret(ExprId),
}

#[derive(Debug, Eq, PartialEq)]
pub enum Type {
    Boolean,
    Function(Vec<TypeId>, TypeId),
    Struct(StructId),
    Number,

    /// This value is used when the statement/expression does not have any value. This is the
    /// case for `let` and `let fn`.
    Void,

    /// This value is used when the statement/expression does not return any value. This is the
    /// case for `ret`.
    None,
}

#[cfg(test)]
mod tests {
    use crate::make_mir;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_main_with_free_statements() {
        let ast = transmute_ast::parse("let main() {}; let a = 1;").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_free_statements_without_main() {
        let ast = transmute_ast::parse("let a = 1;").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_main_without_free_statements() {
        let ast = transmute_ast::parse("let main() {}").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_function_non_mut_param() {
        let ast = transmute_ast::parse("let f(a: number) {}").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_function_mut_param() {
        let ast = transmute_ast::parse("let f(a: number) { a = a + 1; }").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_function_local() {
        let ast = transmute_ast::parse("let f() { let a = 0; }").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_function_mut_local() {
        let ast = transmute_ast::parse("let f() { let a = 0; a = 1; }").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }

    #[test]
    fn test_function_mut_local_shadow_param() {
        let ast = transmute_ast::parse("let f(a: number) { let a = 0; a = 1; }").unwrap();
        let hir = transmute_hir::resolve(ast).unwrap();
        assert_debug_snapshot!(make_mir(hir));
    }
}
