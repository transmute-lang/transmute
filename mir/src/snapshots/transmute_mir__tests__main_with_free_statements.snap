---
source: mir/src/lib.rs
expression: make_mir(hir)
---
Err(
    Diagnostics {
        diagnostics: [
            Diagnostic {
                message: "only functions are allowed at top level",
                span: 1:16; [15, 25],
                level: Error,
                generated_at: (
                    "mir/src/lib.rs",
                    73,
                ),
            },
        ],
    },
)
